django/db/migrations/operations/models.py

```diff
@@ -407,21 +407,37 @@ def database_forwards(self, app_label, schema_editor, from_state, to_state):
                     or not new_field.remote_field.through._meta.auto_created
                 ):
                     continue
-                # Rename the M2M table that's based on this model's name.
+
                 old_m2m_model = old_field.remote_field.through
                 new_m2m_model = new_field.remote_field.through
-                schema_editor.alter_db_table(
-                    new_m2m_model,
-                    old_m2m_model._meta.db_table,
-                    new_m2m_model._meta.db_table,
-                )
-                # Rename the column in the M2M table that's based on this
-                # model's name.
-                schema_editor.alter_field(
-                    new_m2m_model,
-                    old_m2m_model._meta.get_field(old_model._meta.model_name),
-                    new_m2m_model._meta.get_field(new_model._meta.model_name),
-                )
+
+                # Rename the both columns in the M2M table in case of same model names.
+                if (
+                    old_field.related_model._meta.model_name
+                    == old_model._meta.model_name
+                    or new_field.related_model._meta.model_name
+                    == new_model._meta.model_name
+                ):
+                    schema_editor._alter_many_to_many(
+                        new_m2m_model,
```

Comment:
This should be model owning the many-to-many field, not the intermediary auto-created model itself. Refer to how `alter_field` calls `_alter_many_to_many` internally.

https://github.com/django/django/blob/b6ed389eec3e72d50301f456d711496331abd7b3/django/db/backends/base/schema.py#L800-L810